### DISCUSSION TITLE
Setting up a custom OE4T-based distro

### DISCUSSION THREAD

**Discussion by vinotion** [OE4T/tegra-demo-distro#171] [Q&A] [answered]
Hi!

I'm here on behalf of ViNotion, a computer vision company in the Netherlands. We've been pointed at this great project here by Sighthound; thanks for that!

We're currently running our products using a fully customized JetPack-L4T-derived flash image, which has been OK for a couple of years, but we're constantly finding Nvidia changing stuff from underneath us (requiring lots of reverse engineering), and we have no A/B redundancy or OTA upgrading solution for this image yet. The Yocto and Mender-based OE4T distro is just what we've been looking for, and the quality of the scripts and configurations is very high. We're currently rebasing our deployment flash image to OE4T, and in the future we'll gladly contribute to this project.

A broad question that I have: as an 'MVP' we built the tegra-demo-distro and modified it to practice adding layers and dependencies, adding overlay FS paths, changing partition setup etc. -- this is working fine now and was very educational. But for a next step, we'd like to make a production setup where the layers and distro modifications live in their own proper place. Currently, apart from the additional layers, some modifications live in the `meta-tegrademo` layer which, on the long run, seems like the wrong way to do things.

Also, the fact that this project is called `tegra-demo-distro` makes us feel uncomfortable in the sense that it may be changed at any moment in time "because it's just a demo". That doesn't feel like a steady base. As far as we could make up, the `meta-tegrademo` layer and distro setup is pretty custom-made. But perhaps we're missing some Yocto documentation.

Does anyone have any suggestions or pointers as to where to look for how to practically setup a custom production-grade distribution on top of this one? Thanks in advance! 😃 

Best regards,

Kris

**Comment by madisongh** [comment, marked as answer]
The demo distro is, as you say, "just a demo."  You could certainly use it as a starting point for putting together your own distro, and the goal in providing it is to show an example of how to do that.  I definitely wouldn't treat it as an authoritative upstream base, though.  Clone it, test with it, and feel free to reuse whatever parts you find useful when constructing a distro for your own products, but definitely make that distro your own. :)  And if you find, as you are productizing, additions or changes that you think others will find useful, please think about contributing those back here.

**Reply by vinotion** [reply]
OK, that's clear, thanks very much for the answer. Especially the `overlayfs` scripting is useful for cherry-picking stuff for persistent configurations, we're going to keep that for sure ;-)

**Comment by dwalkes** [comment]
Welcome to the party @vinotion !

> Does anyone have any suggestions or pointers as to where to look for how to practically setup a custom production-grade distribution on top of this one?

See https://github.com/OE4T/tegra-demo-distro/wiki/Creating-your-Distro-Layer for some suggestions if you haven't seen this already.

**Reply by vinotion** [reply]
Thanks Dan! We're glad to be here 😃

We've already created a couple of layers for our application firmware and some dependencies, this is working now. We're still figuring out the partitioning now, but we'll get there.
